### PR TITLE
Wrap OANet verifier

### PR DESCRIPTION
**Pull request by johnwlambert**
See: https://arxiv.org/abs/1908.04964

### CODE REVIEW

File: gtsfm/frontend/verifier/oanet.py [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+
+        Args:
+            kpts: array of shape (N,2) representing **unnormalized** 2d keypoint coordinates.
+
+        Returns:
+            nkpts: array of shape (N,2) representing **normalized** 2d keypoint coordinates.
+        """
+        x_mean = np.mean(kpts, axis=0)
+        dist = kpts - x_mean
+        meandist = np.linalg.norm(dist, axis=1).mean()
```
Comment (akshay-krishnan) [inline gtsfm/frontend/verifier/oanet.py:181]: what is meandist?

**Review by akshay-krishnan — APPROVED** [review]

File: gtsfm/frontend/verifier/oanet.py (hunk 1 above)
Comment (johnwlambert) [reply gtsfm/frontend/verifier/oanet.py:181]: It's the average norm of every keypoint coordinate (x,y).

File: gtsfm/frontend/verifier/oanet.py [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+        self._min_matches = (
+            verifier_base.NUM_MATCHES_REQ_E_MATRIX
+            if self._use_intrinsics_in_verification
+            else verifier_base.NUM_MATCHES_REQ_F_MATRIX
+        )
+
+        # for failure, i2Ri1 = None, and i2Ui1 = None, and no verified correspondences, and inlier_ratio_est_model = 0
+        self._failure_result = (None, None, np.array([], dtype=np.uint64), 0.0)
+
+        self.is_cuda = torch.cuda.is_available()
```
Comment (ayushbaid) [inline gtsfm/frontend/verifier/oanet.py:65]: Nit: all class attributes here and below can be private (with the underscore suffix)

File: gtsfm/frontend/verifier/oanet.py [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+        self.default_config["use_ratio"] = 0
+        self.default_config["use_mutual"] = 0  # TODO: add option to use ratio and mutual as side info.
+        self.default_config["iter_num"] = 1
+        self.default_config["inlier_threshold"] = 1
+        self.default_config_ = SimpleNamespace(**self.default_config)
+
+        self.model = OANet(self.default_config_, self.device)
+
+        checkpoint = torch.load(self.model_path, map_location=self.device)
+        self.model.load_state_dict(checkpoint["state_dict"])
```
Comment (ayushbaid) [inline gtsfm/frontend/verifier/oanet.py:83]: Loading the model in the constructor can be very expensive for Dask

File: gtsfm/frontend/verifier/oanet.py [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+
+        self.default_config = {}
+        self.default_config["net_channels"] = 128
+        self.default_config["net_depth"] = 12
+        self.default_config["clusters"] = 500
+        self.default_config["use_ratio"] = 0
+        self.default_config["use_mutual"] = 0  # TODO: add option to use ratio and mutual as side info.
+        self.default_config["iter_num"] = 1
+        self.default_config["inlier_threshold"] = 1
+        self.default_config_ = SimpleNamespace(**self.default_config)
```
Comment (ayushbaid) [inline gtsfm/frontend/verifier/oanet.py:78]: Can we make DEFAULT_CONFIG a global variable? To avoid magic numbers

File: gtsfm/frontend/verifier/oanet.py [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+        if self.is_cuda:
+            self.model = self.model.cuda()
+
+        self.model.eval()
+
+        # Run RANSAC after OANet, for additional verification.
+        self.ransac = Ransac(
+            use_intrinsics_in_verification=use_intrinsics_in_verification,
+            estimation_threshold_px=estimation_threshold_px,
+        )
```
Comment (ayushbaid) [inline gtsfm/frontend/verifier/oanet.py:94]: Accept the additional verifier as an argument?

File: gtsfm/frontend/verifier/oanet.py [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+            camera_intrinsics_i2: intrinsics for image #i2.
+
+        Returns:
+            Estimated rotation i2Ri1, or None if it cannot be estimated.
+            Estimated unit translation i2Ui1, or None if it cannot be estimated.
+            Indices of verified correspondences, of shape (N, 2) with N <= N3. These are subset of match_indices.
+            Inlier ratio w.r.t. the estimated model, i.e. #ransac inliers / # putative matches.
+        """
+        if match_indices.shape[0] < self._min_matches:
+            logger.info("[OANet] Not enough correspondences for verification.")
```
Comment (ayushbaid) [inline gtsfm/frontend/verifier/oanet.py:123]: No need of this tags as I have fixed the logging in the hilti PR.

File: gtsfm/frontend/verifier/oanet.py [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+
+        # only matched keypoints are fed into the model.
+        matched_keypoints_i1 = keypoints_i1.coordinates[match_indices[:, 0]]
+        matched_keypoints_i2 = keypoints_i2.coordinates[match_indices[:, 1]]
+
+        # using OANet's custom normalization utility, obtain two (N,2) arrays
+        normalized_keypoints = [
+            torch.from_numpy(self.__normalize_kpts(matched_keypoints_i1).astype(np.float32)).to(self.device),
+            torch.from_numpy(self.__normalize_kpts(matched_keypoints_i2).astype(np.float32)).to(self.device),
+        ]
```
Comment (ayushbaid) [inline gtsfm/frontend/verifier/oanet.py:137]: Any way we can use calibration to normalize the points? To get rid of distortion for example? Maybe we can do calibration's normalization followed by OANet's normalization

**Review by ayushbaid — APPROVED** [review]
LGTM overall. Please check the test failures

File: gtsfm/frontend/verifier/oanet.py (hunk 4 above)
Comment (johnwlambert) [reply gtsfm/frontend/verifier/oanet.py:78]: Sure, thanks, updated now.

File: gtsfm/frontend/verifier/oanet.py (hunk 3 above)
Comment (ayushbaid) [reply gtsfm/frontend/verifier/oanet.py]: What do you think about this? @johnwlambert

**Comment by akshay-krishnan** [comment]
rebase and merge?

File: gtsfm/frontend/verifier/oanet.py [hunk 8]
```diff
@@ -0,0 +1,188 @@
...
+DEFAULT_CONFIG_DICT["clusters"] = 500
+# Note: using ratio test and mutual matching side info is not possible w/ current GTSFM front-end API.
+DEFAULT_CONFIG_DICT["use_ratio"] = 0
+DEFAULT_CONFIG_DICT["use_mutual"] = 0
+DEFAULT_CONFIG_DICT["iter_num"] = 1
+DEFAULT_CONFIG_DICT["inlier_threshold"] = 1
+DEFAULT_CONFIG = SimpleNamespace(DEFAULT_CONFIG_DICT)
+
+
+class OANetVerifier(VerifierBase):
```
Comment (akshay-krishnan) [inline gtsfm/frontend/verifier/oanet.py:47]: Maybe document when this works well? Only with SIFT?
